corehq/apps/es/users.py

```diff
@@ -68,7 +68,7 @@ def show_only_inactive(self):
 
 class ElasticUser(ElasticDocumentAdapter):
 
-    _index_name = "hqusers_2017-09-07"
+    _index_name = "hqusers_2022-05-09"
```

Comment:
How do you plan on performing this migration? IIRC, this will cause the existing index to be "orphaned" (no longer referenced by HQ, but still present, requiring manual deletion) and the new index to be populated from authoritative data in Couch/SQL (I _think_ this happens during the deploy before Django starts, but I'm not sure).

Pinging @orangejenny here because I recall there being a problem "migrating" this index in the past. It might not be safe to use the "discard + build from scratch" technique on this particular index, in which case a "copy from existing reindex" would be required.

Reply:
+1 to writing a rollout plan for this change

@joxl I think the issue you're thinking of is that the `UnknownUsersPillow` populates this index, but operates on form submissions, so reindexing the users pillow would lose those unknown users.

Reply:
@esoergel, that sounds familiar. Do you know if there are repercussions if we lose those?

Reply:
That's a good question. My thought was to reach out to some people once this PR is approved and ask how to go about getting this safely on prod.

Reply:
Definitely agree with those that say we need a rollout plan, and the sooner we start on that the better. I don't think we need to wait for this to be approved to start discussing that, especially since the merge on this is blocked until rollout is ready.

As far as the specifics of the plan, due to the unkown user issue mentioned in this thread, I don't think it is feasible for us to reindex this from primary data (would require iterating over all forms). 

However, I am not convinced we need a new index at all. If we break this into a two stage process, where the first PR adds the new field and modified existing usages such that they are unaffected by partial data in that field (make sure existing queries ignore this field), we can then reindex all the web user docs (we shouldn't even need to worry about mobile users, right?), and when that finishes, change the code to properly utilize the new data. I think that is just removing the unused `AND` from the first PR, and then adding it back when we are done.

Since unknown users don't have assigned location ids, we won't need to worry about reindexing those docs, just web users, which I suspect are a small portion of overall users in the system, making that process relatively quick. We can probably write up a custom management command to do that (or if we are feeling lazy can just use `ptop_reindexer_V2` and give up the speed bonus of skipping commcare users). 

Either way we will have to add something to the change feed for localhosters to run, and wait an appropriate amount of time after for the final merge.

Reply:
I like that idea. For localhosters, I think this could easily just be run automatically as part of a deploy since none of them have large numbers of users.

Reply:
@calellowitz

 > However, I am not convinced we need a new index at all. If we break this into a two stage process, where the first PR adds the new field

By saying "adds the new field", are you referring to the new field in `user_mapping.py`? If so, why don't we need a new index then, since [this](https://commcare-hq.readthedocs.io/elasticsearch.html?highlight=elasticsearch#adding-data-to-an-index) document specifies that in order to add new data to an index we need to build a new index? Or would `ptop_reindexer_v2` simply be enough to apply a mapping change?


> we can then reindex all the web user docs

How would we specifically reindex only the webuser docs? Aren't we storing all users in the user index (meaning that if a reindex should happen it will happen for all data in that index, thus for all users, be it web user or not)? 

(PS. this is my first time working so intimately with CCHQ's integration with ES, so please bear with me)

Reply:
> By saying "adds the new field", are you referring to the new field in user_mapping.py? If so, why don't we need a new index then, since [this](https://commcare-hq.readthedocs.io/elasticsearch.html?highlight=elasticsearch#adding-data-to-an-index) document specifies that in order to add new data to an index we need to build a new index?

It may be helpful to think of this as analogous to a sql table. Adding that field to the mapping is like adding a column. You can make a whole new table to do that (which is what would happen if you changed the name), or you can just have that new column but with none of the data filled in yet. Then from that point, we can work to fill in the data by updating each web user in the index in place to include the new field.

> How would we specifically reindex only the webuser docs? 

There are a few ways to do this, but one version in pseudocode could be
```
for web_user in get_all_web_users():
    web_user.save()
```
That would trigger the doc to be reprocessed by the pillow and would then include the new fields.There are a few other ways to do that (including skipping the save and publishing directly to kafka, or even sending it to elasticsearch directly), but the specifics are less important than the general idea.

> Or would ptop_reindexer_v2 simply be enough to apply a mapping change?

We could use `ptop_reindexer_v2`, but that would likely be a lot slower because it would include all the mobile users which are not relevant for this change. It would be fine to use though, if you find that easier. We would still need to update the mapping, that however can be done separately from the data migration.

>Aren't we storing all users in the user index (meaning that if a reindex should happen it will happen for all data in that index, thus for all users, be it web user or not)?

Exactly, and that is why I am recommending against a full reindex, rather than just manually resaving web users.

Reply:
@calellowitz Thanks for clarifying this! 

If a new index isn't necessary, I'm more than happy to take another route. I think what you suggested sounds like a simple way to go, so I'll work on that.

Reply:
@Charl1996, if you look at PR #31510 for how to update the mapping. Specifically [this migration file](https://github.com/dimagi/commcare-hq/pull/31510/files#diff-c339acf16ac1825f055724a91c8eac1c424b94b3d76c90c5ee1f18caddaae3b7) demonstrates how to use existing migration tools to perform the work.  The `corehq/ex-submodules/pillowtop/migrations` migrations directory is conventionally where we've started putting these mapping update migrations, and you're free to overwrite migration `0007` because that PR is stalled for now and needs more work before it can continue.